**Оцінювання даної роботи буде ураховувати якість Вашого коду та текстових відповідей. При наявності ЛИШЕ коду , робота оцінюватися НЕ БУДЕ.**  

**Оформлення текстових відповідей повинно проводитися в текстових комірках ноутбука з використанням синтаксису [LaTeX](https://en.wikibooks.org/wiki/LaTeX/Basics) для оформлення тексту та формул.**

**Роботи ідентифіковані як ІДЕНТИЧНІ одразу отримують 0 балів.**

**При реалізації завдань потрібно використовувати методи з бібліотеки numpy та pandas.**

**Рішення з циклами будуть оцінені в 0 балів.**

# I. NumPy.

0. Імпортувати numpy загальноприйнятим способом.

In [793]:
import numpy as np
from faker import Faker

1. Визначити список з оцінками студента(ки) за семестр. Використовуючи його, створити одновимірний масив numpy.

In [794]:
student_marks = np.array([87, 23, 83, 12])

student_marks

array([87, 23, 83, 12])

2. Визначити список зі списками, де кожен рядок списку формально означає магазин, а кожна колонка - певний продукт (батон, яйця першої категорії, молоко та ковбаса), а елементами на пересічі є ціни на ці продукти відповідно. Використовуючи цей список, створити матрицю (двовимірний масив) numpy.

In [795]:
market = np.array([
    [50, 100, 70, 120], 
    [47, 93, 75, 101], 
    [45, 106, 71, 99], 
])

market

array([[ 50, 100,  70, 120],
       [ 47,  93,  75, 101],
       [ 45, 106,  71,  99]])

3. Отримайте тип даних значень з масивів завдань 2 та 3. Виведіть їх на екран.

In [796]:
print(market.dtype)
print(student_marks.dtype)

int64
int64


4. Отримайте форми (кортеж з кількістю рядків та колонок) масивів завдань 2 та 3. Виведіть їх на екран.

In [797]:
print(market.shape)
print(student_marks.shape)

(3, 4)
(4,)


5. Створіть одновимірний масив за допомогою спеціальної функції numpy, який би відображав динаміку стабільно зростаючого (з рівними проміжками) прибутку з продажів  за тиждень, де у перший день не було продажів, а в останній день тижня вдалося заробити 1000 грн 50 коп.

In [798]:
weekly_dynamic_profit = np.linspace(0, 1000.50, 7)

weekly_dynamic_profit

array([   0.  ,  166.75,  333.5 ,  500.25,  667.  ,  833.75, 1000.5 ])

6. Створіть два масиви. Використовуючи їх, продемонструйте відмінність вертикального та горизонтального обʼєднання масивів.

In [799]:
array1 = np.array([
    [1,2,3],
    [4,5,6],
    [7,8,9]
])

array2 = np.array([
    [10,11,12],
    [13,14,15],
    [16,17,18]
])

vertical_merge = np.vstack((array1, array2))
horisontal_merge = np.hstack((array1, array2))

print(vertical_merge, "\n")
print(horisontal_merge)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]
 [13 14 15]
 [16 17 18]] 

[[ 1  2  3 10 11 12]
 [ 4  5  6 13 14 15]
 [ 7  8  9 16 17 18]]


7. Визначити функцію, яка приймає на вході масив і транспонує його. Скористайтесь методом reshape. І не забудьте почати з docstrings.

In [800]:
def transpose_array(array):
    """
    Transpose matrix

    Parameters:
    array (numpy.ndarray): The input matrix to be transposed.

    Returns:
    numpy.ndarray: The transposed array.
    """

    rows = array.shape[0]
    cols = array.shape[1]

    return np.reshape(array, (cols, rows))

array = np.array([
    [1,2,3,4],
    [5,6,7,8],
    [9,10,11,12],
])

reshaped_array = transpose_array(array)

print(array, "\n")
print(reshaped_array)


[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]] 

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


8. Створіть два масиви. Використовучи ці масиви, продемонструйте 1) операцію по-елементного додавання, 2) операцію по-елементного віднімання, 3) множення масиву на число, 4) операцію по-елементного множення, і 5) матричного множення.

In [801]:
array1 = np.array([1, 2, 3])
array2 = np.array([4, 5, 6])

add = array1 + array2
sub = array1 - array2
scalar_mult = array1 * 5
el_mult = array1 * array2
matrix_mult = np.dot(array1, array2.T)

print(add)
print(sub)
print(scalar_mult)
print(el_mult)
print(matrix_mult)

[5 7 9]
[-3 -3 -3]
[ 5 10 15]
[ 4 10 18]
32


9. Створіть двовимірний масив (матрицю) та розрахуйте: 1) мінімальне число, 2) максимальне число, 3) суму чисел, 4) мінімальні числа для кожного рядка, 5) максимальні числа для кожного стовпчика.

In [802]:
array = np.array([
    [1,2,3],
    [4,5,6]
])

print(array.min())
print(array.max())
print(array.sum())
print(array.min(axis=1))
print(array.max(axis=0))

1
6
21
[1 4]
[4 5 6]


10. Створіть матрицю , де деякі елементи повторюються. Знайдіть унікальні значення елементів цієї матриці та їхні частоти.

In [803]:
array = np.array([1, 2, 3, 4, 1, 2, 3, 4, 1, 2])

print(np.unique(array, return_counts = True)[0])
print(np.unique(array, return_counts = True)[1])

[1 2 3 4]
[3 3 2 2]


# II. Importing. Creating DF

0. Імпортувати бібліотеку pandas

In [804]:
import pandas as pd

11. Створити тестовий датафрейм зі списку, що містить записи пацієнтів про їхні прізвища, імена, дату народження, масу тіла та наявність медичного страхування. Одне зі значень зробить пустим. Ознайомтесь з типом кожної колонки.

In [805]:
fake = Faker()

columns = ["Last Name", "First Name", "Date of Birth", "Weight", "Has Insurance"]

data = []
for row in range(10):
    last_name = fake.last_name()
    first_name = fake.first_name()
    date_of_birth = fake.date_of_birth(minimum_age=18, maximum_age=90).strftime('%Y-%m-%d')
    weight = np.nan if row == 9 else np.random.randint(50, 100)
    has_insurance = fake.boolean()
    data.append([last_name, first_name, date_of_birth, weight, has_insurance])

test_df = pd.DataFrame(data, columns=columns)

test_df["Date of Birth"] = pd.to_datetime(test_df["Date of Birth"])
test_df["Weight"] = pd.to_numeric(test_df["Weight"], errors="coerce")
test_df["Has Insurance"] = test_df["Has Insurance"].astype(bool)

print(test_df)

   Last Name First Name Date of Birth  Weight  Has Insurance
0       Hays       Emma    2001-04-14    57.0           True
1      Jones     Daniel    2002-08-31    62.0          False
2       Wood     Meghan    1944-02-17    98.0           True
3     Medina  Elizabeth    1955-08-17    79.0           True
4       Boyd     Joshua    1956-09-09    62.0           True
5    Elliott     Daniel    2006-07-22    89.0           True
6  Gutierrez     Donald    1969-03-08    82.0          False
7      Barry      Shaun    1977-09-26    86.0           True
8    Goodman     Cheryl    1974-11-17    51.0           True
9    Ramirez      Brian    1950-05-05     NaN          False


# III. Reading from file

12. Завантажити файл missile_attacks_daily.csv з [Kaggle](https://www.kaggle.com/datasets/piterfm/massive-missile-attacks-on-ukraine?select=missile_attacks_daily.csv). Ознайомитись з описом датасету. Додати файл до папки поточної сесії colab notebook на панелі ліворуч або на гугл диск (локально до проєкту, якщо працюєте у пайчармі). Зчитати датафрейм з файлу та вивести датафрейм на екран. Зверніть увагу на те, щоб всі колонки відповідали наведеним у описі типам даних.

In [806]:
df = pd.read_csv("./datasets/missile_attacks_daily.csv")

In [807]:
print(df)
print(df.dtypes)

            time_start          time_end               model  \
0     2025-01-18 20:00  2025-01-19 09:00      Shahed-136/131   
1     2025-01-17 19:00  2025-01-18 08:00      Shahed-136/131   
2     2025-01-16 19:00  2025-01-17 09:00      Shahed-136/131   
3     2025-01-16 19:00  2025-01-17 09:00    Iskander-M/KN-23   
4     2025-01-16 19:00  2025-01-17 09:00          Iskander-M   
...                ...               ...                 ...   
2039  2022-10-01 12:00  2022-10-01 12:00            Orlan-10   
2040        2022-09-30        2022-09-30            Orlan-10   
2041  2022-09-29 23:00  2022-09-29 23:59      Shahed-136/131   
2042        2022-09-28        2022-09-28                X-59   
2043        2022-09-28        2022-09-28  Reconnaissance UAV   

                                           launch_place  \
0         Bryansk oblast and Oryol oblast and Millerovo   
1     Primorsko-Akhtarsk and Bryansk oblast and Mill...   
2     Primorsko-Akhtarsk and Kursk oblast and Bryans..

# IV. Data understanding

13. Виведіть на екран перші 10 рядків зчитаного датафрейму.

In [808]:
df.head(10)

time_start          time_end             model  \
0  2025-01-18 20:00  2025-01-19 09:00    Shahed-136/131   
1  2025-01-17 19:00  2025-01-18 08:00    Shahed-136/131   
2  2025-01-16 19:00  2025-01-17 09:00    Shahed-136/131   
3  2025-01-16 19:00  2025-01-17 09:00  Iskander-M/KN-23   
4  2025-01-16 19:00  2025-01-17 09:00        Iskander-M   
5  2025-01-15 20:00  2025-01-16 09:00    Shahed-136/131   
6  2025-01-15 06:00  2025-01-15 08:00         X-59/X-69   
7  2025-01-15 06:00  2025-01-15 08:00         X-22/X-32   
8  2025-01-15 06:00  2025-01-15 08:00       X-101/X-555   
9  2025-01-15 06:00  2025-01-15 08:00            Kalibr   

                                        launch_place       target  carrier  \
0      Bryansk oblast and Oryol oblast and Millerovo      Ukraine      NaN   
1  Primorsko-Akhtarsk and Bryansk oblast and Mill...      Ukraine      NaN   
2  Primorsko-Akhtarsk and Kursk oblast and Bryans...      Ukraine      NaN   
3                 Bryansk oblast and Voronezh oblast  Kyiv oblast      NaN   
4                                      Rostov oblast      Ukraine      NaN   
5  Primorsko-Akhtarsk and Bryansk oblast and Oryo...      Ukraine      NaN   
6                                    Belgorod oblast      Ukraine      NaN   
7                                        Tula oblast      Ukraine  Tu-22M3   
8                                   Volgograd oblast      Ukraine  Tu-95MS   
9                                          Black Sea      Ukraine     Navi   

   launched  destroyed  not_reach_goal  cross_border_belarus  back_russia  \
0      61.0         43            15.0                   NaN          NaN   
1      39.0         24            14.0                   NaN          NaN   
2      50.0         33             9.0                   NaN          NaN   
3       4.0          2             NaN                   NaN          NaN   
4       2.0          0             NaN                   NaN          NaN   
5      55.0         34            18.0                   NaN          NaN   
6       4.0          4             NaN                   NaN          NaN   
7       7.0          0             NaN                   NaN          NaN   
8      27.0         23             NaN                   NaN          NaN   
9       4.0          3             NaN                   NaN          NaN   

   still_attacking                                  destroyed_details  \
0              1.0  {'Mykolaiv oblast': NaN,    'Kyiv oblast': NaN...   
1              NaN  {'Kherson oblast': NaN, 'Khmelnytskyi oblast':...   
2              NaN  {'Odesa oblast and Mykolaiv oblast': 7, 'Khmel...   
3              NaN                                                NaN   
4              NaN                                                NaN   
5              NaN  {'Mykolaiv oblast': 5, 'Kyiv oblast': NaN, 'Su...   
6              NaN                                                NaN   
7              NaN                                                NaN   
8              NaN                                       {'west': 20}   
9              NaN                                       {'south': 1}   

  launched_details launch_place_details  \
0              NaN                  NaN   
1              NaN                  NaN   
2              NaN                  NaN   
3              NaN                  NaN   
4              NaN                  NaN   
5              NaN                  NaN   
6              NaN                  NaN   
7              NaN                  NaN   
8              NaN                  NaN   
9              NaN                  NaN   

                                              source  
0  kpszsu/posts/pfbid0meYNAsX4voALE4SXunVpzKTwzrJ...  
1  kpszsu/posts/pfbid0Rpzfu7u9ERu4o49Gm12eqSogKvZ...  
2  kpszsu/posts/pfbid0Gvye9Wn2gj8RPiiGaTQFGRas9Yg...  
3  kpszsu/posts/pfbid0Rpzfu7u9ERu4o49Gm12eqSogKvZ...  
4  kpszsu/posts/pfbid0Gvye9Wn2gj8RPiiGaTQFGRas9Yg...  
5  kpszsu/posts/pfbid0tPwzQHvvr63PbehDshED2W

14. Отримайте інформацію про форму датафрейму.

In [809]:
df.shape

(2044, 16)

15. Отримайте дані про кожну колонку з її назвою, типом даних та кількістю непустих значень.

In [810]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2044 entries, 0 to 2043
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   time_start            2044 non-null   object 
 1   time_end              2044 non-null   object 
 2   model                 2044 non-null   object 
 3   launch_place          1023 non-null   object 
 4   target                2016 non-null   object 
 5   carrier               225 non-null    object 
 6   launched              2041 non-null   float64
 7   destroyed             2044 non-null   int64  
 8   not_reach_goal        157 non-null    float64
 9   cross_border_belarus  32 non-null     float64
 10  back_russia           34 non-null     float64
 11  still_attacking       28 non-null     float64
 12  destroyed_details     294 non-null    object 
 13  launched_details      4 non-null      object 
 14  launch_place_details  4 non-null      object 
 15  source               

16. Отримайте статистику для колонок числового типу.

In [811]:
print(df["not_reach_goal"].min())
print(df["launched"].max())
print(df["back_russia"].mean())

1.0
193.0
2.0


17. Отримати кількість унікальних значень у кожній колонці. Вивести на екран значення та кількість цих значень у колонці, яка має найменше унікальних значень.

In [812]:
df.nunique()

time_start               864
time_end                1015
model                     61
launch_place             135
target                    97
carrier                   58
launched                 113
destroyed                 81
not_reach_goal            55
cross_border_belarus       7
back_russia                5
still_attacking            8
destroyed_details        290
launched_details           4
launch_place_details       4
source                   960
dtype: int64

In [813]:
df.nunique().min()

np.int64(4)

18. На основі отриманої інформації про датасет вище, опишіть по пунктах, що б ви виправили або на що б звернули увагу на кроці очищення даних і чому?

**місце для текстової відповіді**

1.   convert all colums to right datatype
2.   remove all duplicates
3.   remove all imposible values

# V. Data Cleaning

19. Перетворити колонку launched у цілочислену (Int64). Перевірити, чи все пройшло, як очікувалось.

In [814]:
try:
    df["launched"] = df["launched"].astype(int)
except Exception as error:
    print(error)

Cannot convert non-finite values (NA or inf) to integer


In [815]:
df["launched"].dtype

dtype('float64')

20. Видалити колонки, де кількість непустих значень становить менше 30% від загальної кількості записів.

In [816]:
total_rows = df.shape[0]

deletion_percentage = total_rows * 0.3

df_cleaned: pd.DataFrame = df.dropna(thresh=deletion_percentage, axis=1)

"Original columns:\n", df.columns

('Original columns:\n',
 Index(['time_start', 'time_end', 'model', 'launch_place', 'target', 'carrier',
        'launched', 'destroyed', 'not_reach_goal', 'cross_border_belarus',
        'back_russia', 'still_attacking', 'destroyed_details',
        'launched_details', 'launch_place_details', 'source'],
       dtype='object'))

In [817]:
"Columns after cleaning:\n", df_cleaned.columns

('Columns after cleaning:\n',
 Index(['time_start', 'time_end', 'model', 'launch_place', 'target', 'launched',
        'destroyed', 'source'],
       dtype='object'))

21. Очистити датафрейм від дублікатів (дублюючихся рядків) та записів, де атака продовжувалась відʼємну або нульову кількість часу (що фізично неможливо). Не забудьте оновити індекс. Скільки рядків було видалено? Як ви це перевірили у коді?

In [818]:
df_cleaned['time_start'] = pd.to_datetime(df_cleaned['time_start'], format="mixed", errors='coerce')
df_cleaned['time_end'] = pd.to_datetime(df_cleaned['time_end'], format="mixed", errors='coerce')

df_cleaned = df_cleaned.dropna(subset=['time_start', 'time_end'])

df_cleaned['attack_duration'] = (df_cleaned['time_end'] - df_cleaned['time_start']).dt.total_seconds()

initial_rows = df_cleaned.shape[0]
df_cleaned = df_cleaned[df_cleaned['attack_duration'] > 0]
final_rows = df_cleaned.shape[0]

df_cleaned.reset_index(drop=True, inplace=True)

df_cleaned.duplicated().sum()

C:\Users\Bering\AppData\Local\Temp\ipykernel_14712\2567276174.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['time_start'] = pd.to_datetime(df_cleaned['time_start'], format="mixed", errors='coerce')
C:\Users\Bering\AppData\Local\Temp\ipykernel_14712\2567276174.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['time_end'] = pd.to_datetime(df_cleaned['time_end'], format="mixed", errors='coerce')


np.int64(0)

In [819]:
removed_rows = initial_rows - final_rows
f'Removed {removed_rows} rows'

'Removed 836 rows'

**місце для текстової відповіді**

Dataset has no dublicates and there was removed 836 rows

22. Виведіть на екран всі рядки, де є хоч одне пропущене значення.

In [821]:
df_cleaned.isna().count()

time_start         1208
time_end           1208
model              1208
launch_place       1208
target             1208
launched           1208
destroyed          1208
source             1208
attack_duration    1208
dtype: int64